---
title: "Class_15_Miniproject"
author: "Samuel Do (PID:A15803613)"
date: "3/8/2022"
output: html_document
---
# 1. Investigating pertussis cases by year
```{r}
# [Q1] With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called "cdc" and use ggplot to make a plot of cases numbers over time.
# install.packages("datapasta")
library(datapasta)
```
```{r}
cdc <- data.frame(
                                 Year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L),
         No..Reported.Pertussis.Cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617)
       )

```
```{r}
library(ggplot2)
library(tidyverse)
ggplot(cdc, aes(x=Year, y=No..Reported.Pertussis.Cases)) +
  geom_point() +
  geom_line() +
  labs(title="Pertussis Cases by Year (1922-2019)")
```

# 2. A tale of two vaccines (wP & aP)
```{r}
# [Q2] Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?
ggplot(cdc, aes(x=Year, y=No..Reported.Pertussis.Cases)) +
  geom_point() +
  geom_line() +
  labs(title="Pertussis Cases by Year (1922-2019)") +
  geom_vline(xintercept = 1946, linetype = "dashed", color = "blue") +
  geom_text(aes(x=1946,label="wP", y=2.5e+05), color="blue") +
  geom_vline(xintercept = 1996, linetype = "dashed", color = "red") +
  geom_text(aes(x=1996,label="aP", y=2.5e+05), color="red")
  
```
```{r}
# [Q3] Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?
# This could be due to a newer variant of pertussis bacteria, which arises from bacterial evolution.
```
3. Exploring CMI-PB data
```{r}
# Allows us to read, write and process JSON data
#install.packages("jsonlite")
library("jsonlite")
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
head(subject, 3)
```
```{r}
#[Q4] How may aP and wP infancy vaccinated subjects are in the dataset?
table(subject$infancy_vac)
# There are 47 aP subjects and 49 wP subjects.
#[Q5] How many Male and Female subjects/patients are in the dataset?
table(subject$biological_sex)
# There are 66 female subjects and 30 male subjects.
#[Q6] What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?
table(subject$biological_sex, subject$race)
```
```{r}
library(lubridate)
library(dplyr)
# [Q7] Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?
today()
wP_subject <- subject %>% filter(infancy_vac=="wP")
aP_subject <- subject %>% filter(infancy_vac=="aP")
wP.age <- time_length(today()-ymd(wP_subject$year_of_birth), "years")
aP.age <- time_length(today()-ymd(aP_subject$year_of_birth), "years")
summary(wP.age)
summary(aP.age)
# The average age of the wP subjects was 35.34 years, while the average for aP subjects was 24.50 years. The ages of the two groups of subjects is significantly different as the IQR of the wP subjects' age does not overlap with that of the aP subjects' age.
# [Q8] Determine the age of all individuals at time of boost?
time_length(ymd(subject$date_of_boost)-ymd(subject$year_of_birth), "years")
# [Q9] With the help of a faceted boxplot (see below), do you think these two groups are significantly different?
age <- today()-ymd(subject$year_of_birth)
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) 
```
# Joining multiple tables
```{r}
# Complete the API URLs...
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/ab_titer", simplifyVector = TRUE) 
head(specimen)
```
```{r}
# [Q9] Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)

# [Q10] Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.
abdata <- inner_join(titer, meta)
dim(abdata)
head(abdata)
```
```{r}
# [Q11] How many specimens (i.e. entries in abdata) do we have for each isotype?
table(abdata$isotype)
# There are 6698 IgE isotypes, 1413 IgG isotypes, 6141 IgG1 isotypes, 6141 IgG2 isotypes, 6141 IgG3 isotypes, and 6141 IgG4 isotypes.
# [Q12] What do you notice about the number of visit 8 specimens compared to other visits?
table(abdata$visit)
# There is a very small number of visit 8 specimens compared to the other visits.
```
4. Examine IgG1 Ab titer levels
```{r}
# Filter IgG1 isotypes and exlude visit 8 subjects
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```
```{r}
#[Q13] Complete the following code to make a summary boxplot of Ab titer levels for all antigens:
ggplot(ig1) +
  aes(ab_titer, antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2)
#[Q14] What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?
# Antigen FIM2/3 is likely the antigen that is produced by bacteria that causes pertussis. Therefore, with the introduction of the vaccines, IgG1 antibodies are more likely to recognize these antigens, resulting in more immunity.
```
```{r}
#[Q15] Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).
filter(ig1, antigen=="Measles") %>%
  ggplot() +
  aes(ab_titer, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title="Measles Antigen Levels per Visit (aP red, wP teal)")
filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(ab_titer, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title="FIM2/3 Antigen Levels per Visit (aP red, wP teal)")
#[Q16] What do you notice about these two antigens time course and the FIM2/3 data in particular?
# The FIM2/3 antigen levels increase by a significant amount compared to the measles antigen levels, which stayed relative the same throughout all visits. The FIM2/3 antigen levels peak at visit 5 and decline by visit 6 onwards.
#[Q17] Do you see any clear difference in aP vs. wP responses?
# While there is no statistically significant difference between aP and wP responses due to the overlap in IQRs in all visits, FIM2/3 antigen levels were relatively greater with the aP vaccine after visit 3.
```
5. Obtaining CMI-PB RNASeq data
```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"
rna <- read_json(url, simplifyVector = TRUE) 
#Join rna data with meta data
ssrna <- inner_join(rna, meta)
head(ssrna)
```
```{r}
#[Q18] Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
#[Q19] What do you notice about the expression of this gene (i.e. when is it at its maximum level)?
# Expression of this gene is relatively low and suddenly peaks by visit 4, but then drops by visit 5.
#[Q20] Does this pattern in time match the trend of antibody titer data? If not, why not?
# This pattern does not match the trend of antibody titer data as antibodies as the general expression of the gene would be more consistent and not suddenly peak and drop. This is because antibodies are long-lasting, which is not shown in this plot.
```
```{r}
# Digging deeper and facet by infancy_vac status
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
# No obvious wP vs. aP differences even if focus on one visit
```

