---
title: "Class09"
author: "Samuel Do (PID:A15803613)"
date: "2/15/2022"
output: html_document
---

```{r}
# PDB database is main repository for 3D structure

fml <- read.csv("Data Export Summary.csv", row.names=1) 
fml

# [Q1] What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

method.total <- colSums(fml) 
method.total 
round(method.total/method.total["Total"]*100, 3) 
# Of the structures in the PDB, 87.197% are solved by X-ray while 5.354% are solved by Electron Microscopy

# [Q2] What proportion of structures in the PDB are protein?

round(fml$Total/method.total["Total"]*100, 3)
# 87.270% of the structures in the PDB are proteins

# [Q3] Type HIV in the PDBwebsite search box on the home page and determine how many HIV-1 protease structures are in the current PDB?
# There are 4483 HIV-1 protease structure in the current PDB
```

```{r}
# Visualizing HIV-1 protease structure using VMD

# [Q4] Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?
# The water molecules are represented as only one sphere because of the negligible size of the hydrogen atoms in comparison to the entire protein.

# [Q5] There is a conserved water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have (see note below)?
# The conserved water molecule has a residue number of 308:0. 
```

```{r}
#install bio3d package using install.packages(bio3d)
library(bio3d)

pdb <-read.pdb("1hsg")
pdb
attributes(pdb)
head(pdb$atom)
```

# [Q7] How many amino acid residues are there in this pdb object?
# There are 198 amino acid residues.

# [Q8] Name one of the two non-protein residues?
# HOH (Water) is one of the non-protein residues.

# [Q9] How many protein chains are in this structure?
# There are 2 protein chains in this structure.

# Comparative structure analysis of Adenylate Kinase(ADK)
```{r}
#install.packages("bio3d")
#install.packages("ggplot2")
#install.packages("ggrepel")
#install.packages("devtools")
#install.packages("BiocManager")

#BiocManager::install("msa")
#devtools::install_bitbucket("Grantlab/bio3d-view")
```

# [Q10] Which of the packages above is found only on BioConductor and not CRAN?
# msa

# [Q11] Which of the above packages is not found on BioConductor or CRAN?: 
# bio3d-view

# [Q12] True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket?
# TRUE

```{r}
#Extract sequence for ADK
aa <- get.seq("1ake_A")
aa

```
# [Q13] How many amino acids are in this sequence, i.e. how long is this sequence?
# This sequence is 214 amino acids long.

```{r}
# Blast search of ADK
blast <- blast.pdb(aa)
```

```{r}
hits <- plot(blast)
hits$pdb.id
```
```{r}
# Download related PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```
# Align and superimpose PDB structures
# pdbaln function did not work due to error in muscle.exe download
```{r}

# Align releated PDBs
#pdbs <- pdbaln(files, fit = TRUE)#, exefile="msa")


# Vector containing PDB codes for figure axis
# ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
# plot(pdbs, labels=ids)

```

# PCA of PDB data
```{r}
#pc.xray <- pca(pdbs)
#plot(pc.xray)
```
```{r}
# RMSD calculation
#rd <- rmsd(pdbs)

# Structure-based clustering
#hc.rd <- hclust(dist(rd))
#grps.rd <- cutree(hc.rd, k=3)

#plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

# Normal Mode Analysis

```{r}
#pdb <- read.pdb("1ake")
#chain <- trim.pdb(pdb, chain="A")
#modes <- nma(pdbs) 
#plot(modes, pdbs, col=grps.rd)
```
# [Q14] What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?
# The black and colored lines are similar in some aspects, but they greatly differ at residues 40-60 and 125-150. This difference at these two regions is likely due to the different residues used to form the nucleotide binding sites of ADK.